beaconChain/src/stateTransition/epoch/variables.ts

```diff
@@ -0,0 +1,111 @@
+  // Variables
+  const currentEpoch: Epoch = getCurrentEpoch(state);
+  const previousEpoch: Epoch = getPreviousEpoch(state);
+  const nextEpoch: Epoch = currentEpoch.addn(1);
+
+  // Validators attesting during the current epoch
+  const currentTotalBalance: Gwei = getTotalBalance(state, getActiveValidatorIndices(state.validatorRegistry, currentEpoch));
+  const currentEpochAttestations: PendingAttestation[] = state.latestAttestations.filter((attestation) => currentEpoch === slotToEpoch(attestation.data.slot));
+
+  // Validators justifying the epoch boundary block at the start of the current epoch
+  const currentEpochBoundaryAttestations: PendingAttestation[] = currentEpochAttestations.filter((attestation) => {
+    if (attestation.data.epochBoundaryRoot === getBlockRoot(state, getEpochStartSlot(currentEpoch))) {
+      return attestation;
+    }
+  });
+
+  const currentEpochBoundaryAttesterIndices: ValidatorIndex[] = currentEpochBoundaryAttestations
+    .map((attestation: PendingAttestation) => getAttestationParticipants(state, attestation.data, attestation.aggregationBitfield))
+    .reduce((previousValue: ValidatorIndex[], currentValue: ValidatorIndex[]) => previousValue.concat(...currentValue));
+
+  const currentEpochBoundaryAttestingBalance = getTotalBalance(state, currentEpochBoundaryAttesterIndices);
+
+  // Validators attesting during the previous epoch
+  const previousTotalBalance = getTotalBalance(state, getActiveValidatorIndices(state.validatorRegistry, previousEpoch));
+
+  // Validators that made an attestation during the previous epoch, targeting the previous justified slot
+  const previousEpochAttestations: PendingAttestation[] = state.latestAttestations.filter((attestation) => {
+    if (previousEpoch === slotToEpoch(attestation.data.slot)) {
+      return attestation;
+    }
+  });
+
+  const previousEpochAttesterIndices: ValidatorIndex[] = previousEpochAttestations
+    .map((attestation: PendingAttestation) => getAttestationParticipants(state, attestation.data, attestation.aggregationBitfield))
+    .reduce((previousValue: ValidatorIndex[], currentValue: ValidatorIndex[]) => previousValue.concat(...currentValue));
+
+  const previousEpochAttestingBalance: Gwei = getTotalBalance(state, previousEpochAttesterIndices);
+
+  // Validators justifying the epoch boundary block at the start of the previous epoch
+  const previousEpochBoundaryAttestations: PendingAttestation[] = previousEpochAttestations.filter((attestation) => {
+    if (attestation.data.epochBoundaryRoot === getBlockRoot(state, getEpochStartSlot(previousEpoch))) {
+      return attestation;
+    }
+  });
+
+  const previousEpochBoundaryAttesterIndices: ValidatorIndex[] = previousEpochBoundaryAttestations
+    .map((attestation) => getAttestationParticipants(state, attestation.data, attestation.aggregationBitfield))
+    .reduce((previousValue: ValidatorIndex[], currentValue: ValidatorIndex[]) => previousValue.concat(...currentValue));
```

Comment:
deduplicate
see the comment for `currentEpochBoundaryAttesterIndices` above

Reply:
dunzo